packages/core/src/amazonq/lsp/lspController.ts

```diff
@@ -0,0 +1,361 @@
+/*!
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+import * as vscode from 'vscode'
+import * as path from 'path'
+import * as fs from 'fs-extra'
+import * as crypto from 'crypto'
+import { getLogger } from '../../shared/logger/logger'
+import { CurrentWsFolders, collectFilesForIndex } from '../../shared/utilities/workspaceUtils'
+import fetch from 'node-fetch'
+import { LspClient } from './lspClient'
+import AdmZip from 'adm-zip'
+import { RelevantTextDocument } from '@amzn/codewhisperer-streaming'
+import { makeTemporaryToolkitFolder } from '../../shared/filesystemUtilities'
+import { CodeWhispererSettings } from '../../codewhisperer/util/codewhispererSettings'
+import { activate as activateLsp } from './lspClient'
+import { telemetry } from '../../shared/telemetry'
+import { isCloud9 } from '../../shared/extensionUtilities'
+import { globals } from '../../shared'
+import { AuthUtil } from '../../codewhisperer'
+import { isWeb } from '../../shared/extensionGlobals'
+
+function getProjectPaths() {
+    const workspaceFolders = vscode.workspace.workspaceFolders
+    if (!workspaceFolders || workspaceFolders.length === 0) {
+        throw Error('No workspace folders found')
+    }
+    return workspaceFolders.map(folder => folder.uri.fsPath)
+}
+
+export interface Chunk {
+    readonly filePath: string
+    readonly content: string
+    readonly context?: string
+    readonly relativePath?: string
+    readonly programmingLanguage?: string
+}
+
+export interface Content {
+    filename: string
+    url: string
+    hashes: string[]
+    bytes: number
+    serverVersion?: string
+}
+
+export interface Target {
+    platform: string
+    arch: string
+    contents: Content[]
+}
+
+export interface Manifest {
+    manifestSchemaVersion: string
+    artifactId: string
+    artifactDescription: string
+    isManifestDeprecated: boolean
+    versions: {
+        serverVersion: string
+        isDelisted: boolean
+        targets: Target[]
+    }[]
+}
+const manifestUrl = 'https://aws-toolkit-language-servers.amazonaws.com/temp/manifest.json'
+// this LSP client in Q extension is only going to work with these LSP server versions
+const supportedLspServerVersions = ['0.1.0']
+
+const nodeBinName = process.platform === 'win32' ? 'node.exe' : 'node'
+
+export class LspController {
```

Comment:
what is the purpose of this class? Needs a docstring.  This "controller" object stores no state, so why is it a singleton?

I see "controller" added to many class names without any purpose. It often is just a collection of util functions. If that's the case, why not just have a `lspUtil.ts` module?

Reply:
1. In the latest commit, it does have one state, is Indexing in progress. 

2. Making it a object makes it easier to use sinon.stub for unit testing.

3. It is named as controller because it manages some state ( not having lsp --> install lsp --> lsp running -> indexing in progress)

I will add doc string